## Data cleaning and merging

Before we start with any real analysis, we have to make sure that our dataset is clean and well writte. We want to have only meaningful data. Also, if we lack some data, we would like to merge our dataset with a new one, to enrich the information about the movies.
We start of by importing needed packages.

In [5]:
import pandas as pd
import numpy as np
import tarfile
import zipfile
import re

Then, we import our data, and print a part of it to get a first glance at information we have.

In [4]:
'''
tar_path = 'MovieSummaries.tar.gz'
extract_dir = '/data'

with tarfile.open(tar_path, 'r') as tar:
    tar.extractall(path=extract_dir, filter='data')

print(f'Files extracted to {extract_dir}')
'''

Files extracted to /data


In [6]:
data = pd.read_csv('../../data/MovieSummaries/movie.metadata.tsv', sep='\t', header=None) #here if I don't do header=none, the first row becomes header, but it isn't
data.head()

0           1                                                  2  \
0    975900   /m/03vyhn                                     Ghosts of Mars   
1   3196793   /m/08yl5d  Getting Away with Murder: The JonBenét Ramsey ...   
2  28463795  /m/0crgdbh                                        Brun bitter   
3   9363483  /m/0285_cd                                   White Of The Eye   
4    261236   /m/01mrr1                                  A Woman in Flames   

            3           4      5                                   6  \
0  2001-08-24  14010832.0   98.0  {"/m/02h40lc": "English Language"}   
1  2000-02-16         NaN   95.0  {"/m/02h40lc": "English Language"}   
2        1988         NaN   83.0  {"/m/05f_3": "Norwegian Language"}   
3        1987         NaN  110.0  {"/m/02h40lc": "English Language"}   
4        1983         NaN  106.0   {"/m/04306rv": "German Language"}   

                                           7  \
0  {"/m/09c7w0": "United States of America"}   
1  {"/m/09c7w0": "United States of America"}   
2                     {"/m/05b4w": "Norway"}   
3             {"/m/07ssc": "United Kingdom"}   
4                    {"/m/0345h": "Germany"}   

                                                   8  
0  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
1  {"/m/02n4kr": "Mystery", "/m/03bxz7": "Biograp...  
2  {"/m/0lsxr": "Crime Fiction", "/m/07s9rl0": "D...  
3  {"/m/01jfsb": "Thriller", "/m/0glj9q": "Erotic...  
4                            {"/m/07s9rl0": "Drama"}

In [7]:
# It would be nice to change names of the columns, to know what we are working with
data.columns = ['wikipedia_id','freebase_id','name','release_date','revenue','runtime','languages','countries','genres']
data.head()

wikipedia_id freebase_id  \
0        975900   /m/03vyhn   
1       3196793   /m/08yl5d   
2      28463795  /m/0crgdbh   
3       9363483  /m/0285_cd   
4        261236   /m/01mrr1   

                                                name release_date     revenue  \
0                                     Ghosts of Mars   2001-08-24  14010832.0   
1  Getting Away with Murder: The JonBenét Ramsey ...   2000-02-16         NaN   
2                                        Brun bitter         1988         NaN   
3                                   White Of The Eye         1987         NaN   
4                                  A Woman in Flames         1983         NaN   

   runtime                           languages  \
0     98.0  {"/m/02h40lc": "English Language"}   
1     95.0  {"/m/02h40lc": "English Language"}   
2     83.0  {"/m/05f_3": "Norwegian Language"}   
3    110.0  {"/m/02h40lc": "English Language"}   
4    106.0   {"/m/04306rv": "German Language"}   

                                   countries  \
0  {"/m/09c7w0": "United States of America"}   
1  {"/m/09c7w0": "United States of America"}   
2                     {"/m/05b4w": "Norway"}   
3             {"/m/07ssc": "United Kingdom"}   
4                    {"/m/0345h": "Germany"}   

                                              genres  
0  {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
1  {"/m/02n4kr": "Mystery", "/m/03bxz7": "Biograp...  
2  {"/m/0lsxr": "Crime Fiction", "/m/07s9rl0": "D...  
3  {"/m/01jfsb": "Thriller", "/m/0glj9q": "Erotic...  
4                            {"/m/07s9rl0": "Drama"}

In [8]:
#I would like to see the whole text in genres column :)
pd.set_option('display.max_colwidth', 200)
data.head()

wikipedia_id freebase_id  \
0        975900   /m/03vyhn   
1       3196793   /m/08yl5d   
2      28463795  /m/0crgdbh   
3       9363483  /m/0285_cd   
4        261236   /m/01mrr1   

                                                    name release_date  \
0                                         Ghosts of Mars   2001-08-24   
1  Getting Away with Murder: The JonBenét Ramsey Mystery   2000-02-16   
2                                            Brun bitter         1988   
3                                       White Of The Eye         1987   
4                                      A Woman in Flames         1983   

      revenue  runtime                           languages  \
0  14010832.0     98.0  {"/m/02h40lc": "English Language"}   
1         NaN     95.0  {"/m/02h40lc": "English Language"}   
2         NaN     83.0  {"/m/05f_3": "Norwegian Language"}   
3         NaN    110.0  {"/m/02h40lc": "English Language"}   
4         NaN    106.0   {"/m/04306rv": "German Language"}   

                                   countries  \
0  {"/m/09c7w0": "United States of America"}   
1  {"/m/09c7w0": "United States of America"}   
2                     {"/m/05b4w": "Norway"}   
3             {"/m/07ssc": "United Kingdom"}   
4                    {"/m/0345h": "Germany"}   

                                                                                                                                                                                       genres  
0  {"/m/01jfsb": "Thriller", "/m/06n90": "Science Fiction", "/m/03npn": "Horror", "/m/03k9fj": "Adventure", "/m/0fdjb": "Supernatural", "/m/02kdv5l": "Action", "/m/09zvmj": "Space western"}  
1                                                                              {"/m/02n4kr": "Mystery", "/m/03bxz7": "Biographical film", "/m/07s9rl0": "Drama", "/m/0hj3n01": "Crime Drama"}  
2                                                                                                                                        {"/m/0lsxr": "Crime Fiction", "/m/07s9rl0": "Drama"}  
3                                                                                            {"/m/01jfsb": "Thriller", "/m/0glj9q": "Erotic thriller", "/m/09blyk": "Psychological thriller"}  
4                                                                                                                                                                     {"/m/07s9rl0": "Drama"}

So, in countries, languages and genres, the elements are formated as Freebase ID:name tuples. This is okay, but it would be more readable do have just the names of countries/languages/genres, so we will filter this.

In [9]:
def extract_from_str(extract_str):
    extracted = re.findall(r'":\s*"([^"]+)"', extract_str) #regex for our case
    return ', '.join(extracted) 

data['genres'] = data['genres'].apply(extract_from_str)
data['countries'] = data['countries'].apply(extract_from_str)
data['languages'] = data['languages'].apply(extract_from_str)
data.head()

wikipedia_id freebase_id  \
0        975900   /m/03vyhn   
1       3196793   /m/08yl5d   
2      28463795  /m/0crgdbh   
3       9363483  /m/0285_cd   
4        261236   /m/01mrr1   

                                                    name release_date  \
0                                         Ghosts of Mars   2001-08-24   
1  Getting Away with Murder: The JonBenét Ramsey Mystery   2000-02-16   
2                                            Brun bitter         1988   
3                                       White Of The Eye         1987   
4                                      A Woman in Flames         1983   

      revenue  runtime           languages                 countries  \
0  14010832.0     98.0    English Language  United States of America   
1         NaN     95.0    English Language  United States of America   
2         NaN     83.0  Norwegian Language                    Norway   
3         NaN    110.0    English Language            United Kingdom   
4         NaN    106.0     German Language                   Germany   

                                                                              genres  
0  Thriller, Science Fiction, Horror, Adventure, Supernatural, Action, Space western  
1                                     Mystery, Biographical film, Drama, Crime Drama  
2                                                               Crime Fiction, Drama  
3                                  Thriller, Erotic thriller, Psychological thriller  
4                                                                              Drama

Now that are dataset is readable, we can delve deeper into our data. Are there any missing values? Let's cheeck!

In [10]:
data.replace('', np.nan, inplace=True) #for columns that are strings, we check if there is an empty string
nan_count = data.isna().sum()
print(nan_count)

wikipedia_id        0
freebase_id         0
name                0
release_date     6902
revenue         73340
runtime         20450
languages       13866
countries        8154
genres           2294
dtype: int64


What we can see for some movies is that release date can be problematic - some ovies have just the year, some month and year, and some the full date, while some don't have any date!

In [12]:
exact_date = data['release_date'].str.match(r'^\d{4}-\d{2}-\d{2}$')
month_year = data['release_date'].str.match(r'^\d{4}-\d{2}$') # there are some movies who don't have the full exact date, but do have the month and year
year = data['release_date'].str.match(r'^\d{4}$')

print(f"Number of non valid dates: {data['release_date'].isna().sum()}")
print(f"Number of exact dates: {exact_date.sum()}")
print(f"Number of month-year dates: {month_year.sum()}")
print(f"Number of year dates: {year.sum()}")
print(f"Number of movies: {data.shape[0]}")

Number of non valid dates: 6902
Number of exact dates: 39373
Number of month-year dates: 3294
Number of year dates: 32172
Number of movies: 81741


There is quite a lot of movies that don't have the full date... We can try to get this information from some other dataset
But how do we match movies from different datasets? Is the name enough?

In [13]:
duplicate_names = data['name'].value_counts()[data['name'].value_counts() > 1]

print("Number of duplicate movie names:", duplicate_names.shape[0])
print("Duplicate movie names and their counts:")
print(duplicate_names)

Number of duplicate movie names: 4387
Duplicate movie names and their counts:
name
Alice in Wonderland      17
Macbeth                  16
Les Misérables           13
A Christmas Carol        13
Hero                     12
                         ..
Perfect Day               2
Straw Dogs                2
Falling in Love Again     2
Aradhana                  2
Double Vision             2
Name: count, Length: 4387, dtype: int64


Nope, there are actually some movies that have the same name, exactly 4387 of them. The only other column that could be unique, outside of ids, is the date, so we could check if there are movies that have both the same date as well as the same name.

Just in case, let's first check if there are any full on duplicates in dataset

In [14]:
duplicates = data[data.duplicated(keep=False)]
print(duplicates)

Empty DataFrame
Columns: [wikipedia_id, freebase_id, name, release_date, revenue, runtime, languages, countries, genres]
Index: []


Cool! Now let's compare the dates too!

In [15]:
duplicates_name_date = data[data.duplicated(subset=['name', 'release_date'], keep=False)]
duplicates_name_date.shape[0]

196

196 movies with same name and date, but how many of them actually have a date?

In [16]:
duplicates_name_date_no_nan = duplicates_name_date[duplicates_name_date['release_date'].notna()]
duplicates_name_date_no_nan.shape[0]

72

 Do some of the movies have the same name and year, but one has the exact date?

In [20]:
duplicates_names = data[data.duplicated(subset=['name'], keep=False)]
duplicates_names['release_date'] = pd.to_datetime(duplicates_names['release_date'], errors='coerce')
duplicates_names['release_year'] = duplicates_names['release_date'].dt.year
year_only = duplicates_names[duplicates_names['release_date'].dt.month.isna() & duplicates_names['release_date'].dt.day.isna()]
full_date = duplicates_names[~duplicates_names['release_date'].dt.month.isna() & ~duplicates_names['release_date'].dt.day.isna()]
merged_duplicates = pd.merge(year_only, full_date, on=['name', 'release_year'], suffixes=('_year_only', '_full_date'))

print(merged_duplicates[['name', 'release_year', 'release_date_year_only', 'release_date_full_date']])

Empty DataFrame
Columns: [name, release_year, release_date_year_only, release_date_full_date]
Index: []


C:\Users\Marija\AppData\Local\Temp\ipykernel_17904\1182206444.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  duplicates_names['release_date'] = pd.to_datetime(duplicates_names['release_date'], errors='coerce')
C:\Users\Marija\AppData\Local\Temp\ipykernel_17904\1182206444.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  duplicates_names['release_year'] = duplicates_names['release_date'].dt.year


Okay, at least that sin't happening!

Let's check an example of a movie that has the same name and year, but there are two diffrent movies

In [21]:
data[data['name']=='Intermezzo']

wikipedia_id freebase_id        name release_date  revenue  runtime  \
69072       4599747   /m/0cbxys  Intermezzo         1939      NaN     70.0   
75375       7816833  /m/026f1dz  Intermezzo         1939      NaN     93.0   

                               languages                 countries  \
69072                   English Language  United States of America   
75375  Swedish Language, German Language                    Sweden   

                                                     genres  
69072  Romantic drama, Romance Film, Drama, Black-and-white  
75375                  Romance Film, Drama, Black-and-white

How many of these movies have all the values?

In [22]:
duplicates_name_date_no_nan[duplicates_name_date_no_nan.notna().all(axis=1)]

wikipedia_id freebase_id                  name release_date  \
16826       6061594   /m/0fn46x                Gossip         2000   
80080       3145964   /m/08vczx  Yours, Mine and Ours   2005-11-23   

          revenue  runtime                           languages  \
16826  12591270.0     90.0  Italian Language, English Language   
80080  72028752.0     88.0                    English Language   

                      countries  \
16826  United States of America   
80080  United States of America   

                                                                 genres  
16826  Thriller, Psychological thriller, Mystery, Drama, Suspense, Teen  
80080                                 Romance Film, Family Film, Comedy

Not many... Only 2.

What about the same name same genre?

In [23]:
duplicates_name_genre = data[data.duplicated(subset=['name', 'genres'], keep=False)]
duplicates_name_genre

wikipedia_id freebase_id                     name release_date  \
273        28327554  /m/0cnxf8t             The Gamblers         1970   
392        22697314  /m/05zkfkf         Inherit the Wind   1999-05-29   
640        32505827  /m/0h1cbks                     Dost          NaN   
874        34129990  /m/0hr7b9d              Bella Donna   1915-11-15   
1170       27744173  /m/0cc5j4y            The Long Haul         1988   
...             ...         ...                      ...          ...   
80898       7999199  /m/026m_h7                   Desire         1936   
80956      32998352  /m/0h536dw  A Daughter of Australia         1922   
81159      14373330  /m/03d1zb1   The Brothers Karamazov   1969-07-12   
81295      24162182  /m/07k8ww8            A Million Bid   1927-05-28   
81319      26743619  /m/0bmj7yy             The Calendar   1931-10-31   

       revenue  runtime                           languages  \
273        NaN     93.0                    English Language   
392        NaN    113.0                    English Language   
640        NaN      NaN                      Hindi Language   
874        NaN      NaN       Silent film, English Language   
1170       NaN    113.0                 Portuguese Language   
...        ...      ...                                 ...   
80898      NaN     95.0  English Language, Spanish Language   
80956      NaN      NaN       Silent film, English Language   
81159      NaN    232.0                    Russian Language   
81295      NaN     70.0                    English Language   
81319      NaN     80.0                    English Language   

                      countries  \
273    United States of America   
392    United States of America   
640                       India   
874                         NaN   
1170                     Brazil   
...                         ...   
80898  United States of America   
80956                       NaN   
81159              Soviet Union   
81295  United States of America   
81319            United Kingdom   

                                                       genres  
273                                                     Drama  
392    Period piece, Courtroom Drama, Drama, Television movie  
640                                                       NaN  
874                                                     Drama  
1170                                                    Drama  
...                                                       ...  
80898                                  Drama, Black-and-white  
80956                                             Silent film  
81159                                                   Drama  
81295                        Silent film, Drama, Romance Film  
81319                                                   Drama  

[652 rows x 9 columns]

So, it seems like combining name and date would be our best best to distinct movies. To ensure uniqueness, we are going to delete all duplicates that have the same name and date.

In [24]:
data_no_duplicates = data.drop_duplicates(subset=['name', 'release_date'], keep=False)

Okay, now that we kind of got the feeling of what makes a movie unique, we should check if we have any other odd values in our dataset, aside from nan values. It seems like name, countries, genres and languages could be practically anything. But revenue, if it is not nan, can't really be negative - we should check this.

In [25]:
data_no_duplicates[data_no_duplicates['revenue']<0]['revenue']

Series([], Name: revenue, dtype: float64)

Okay! That's seems fine. Now, we would like to fill in the missing values - we are going to use a new dataset to supllement ours! We will first try with dataset from kaggle: https://www.kaggle.com/datasets/asaniczka/tmdb-movies-dataset-2023-930k-movies?resource=download.

In [26]:
'''
with zipfile.ZipFile('kaggle.zip', 'r') as zip_ref:
    zip_ref.extractall('./data/TMDB_movie/')
'''

In [27]:
data_kaggle = pd.read_csv('../../data/TMDB_movie/TMDB_movie_dataset_v11.csv', sep=',')
data_kaggle.head()

id            title  vote_average  vote_count    status release_date  \
0   27205        Inception         8.364       34495  Released   2010-07-15   
1  157336     Interstellar         8.417       32571  Released   2014-11-05   
2     155  The Dark Knight         8.512       30619  Released   2008-07-16   
3   19995           Avatar         7.573       29815  Released   2009-12-15   
4   24428     The Avengers         7.710       29166  Released   2012-04-25   

      revenue  runtime  adult                     backdrop_path  ...  \
0   825532764      148  False  /8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg  ...   
1   701729206      169  False  /pbrkL804c8yAv3zBZR4QPEafpAR.jpg  ...   
2  1004558444      152  False  /nMKdUUepR0i5zn0y1T4CsSB5chy.jpg  ...   
3  2923706026      162  False  /vL5LR6WdxWPjLPFRLe133jXWsh5.jpg  ...   
4  1518815515      143  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg  ...   

    original_title  \
0        Inception   
1     Interstellar   
2  The Dark Knight   
3           Avatar   
4     The Avengers   

                                                                                                                                                                                                  overview  \
0  Cobb, a skilled thief who commits corporate espionage by infiltrating the subconscious of his targets is offered a chance to regain his old life as payment for a task considered to be impossible: ...   
1   The adventures of a group of explorers who make use of a newly discovered wormhole to surpass the limitations on human space travel and conquer the vast distances involved in an interstellar voyage.   
2  Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attorney Harvey Dent, Batman sets out to dismantle the remaining criminal organizations that plague the s...   
3                          In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.   
4  When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director of the international peacekeeping agency known as S.H.I.E.L.D., finds himself in need of a team to pu...   

  popularity                       poster_path  \
0     83.952  /oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg   
1    140.241  /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg   
2    130.643  /qJ2tW6WMUDux911r6m7haRef0WH.jpg   
3     79.932  /kyeqWdyUXW608qlYkRqosgbbJyK.jpg   
4     98.082   /RYMX2wcKCBAr24UyPD7xwmjaTn.jpg   

                                                      tagline  \
0                        Your mind is the scene of the crime.   
1  Mankind was born on Earth. It was never meant to die here.   
2                           Welcome to a world without rules.   
3                                 Enter the world of Pandora.   
4                                     Some assembly required.   

                                        genres  \
0           Action, Science Fiction, Adventure   
1            Adventure, Drama, Science Fiction   
2               Drama, Action, Crime, Thriller   
3  Action, Adventure, Fantasy, Science Fiction   
4           Science Fiction, Action, Adventure   

                                                              production_companies  \
0                               Legendary Pictures, Syncopy, Warner Bros. Pictures   
1                              Legendary Pictures, Syncopy, Lynda Obst Productions   
2  DC Comics, Legendary Pictures, Syncopy, Isobel Griffiths, Warner Bros. Pictures   
3  Dune Entertainment, Lightstorm Entertainment, 20th Century Fox, Ingenious Media   
4                                                                   Marvel Studios   

                       production_countries  \
0  United Kingdom, United States of America   
1  United Kingdom, United States of America   
2  United Kingdom, United States of America   
3  United States of America, Un

In [28]:
data_kaggle[data_kaggle['revenue']!=0.0].shape[0]
# Oof, not a lot of them have revenues, but still, maybe this dataset does overlap with a lot of ours missing

20759

In [29]:
# What would mess up our work is if this dataset has duplicates, so we should check that
duplicates_kaggle = data_kaggle[data_kaggle.duplicated()]
duplicates_kaggle.shape[0]

368

In [30]:
# It does! We will drop these duplicates
data_kaggle_cleaned = data_kaggle.drop_duplicates()
print(f"Original shape: {data_kaggle.shape[0]}")
print(f"Shape after dropping: {data_kaggle_cleaned.shape[0]}")

Original shape: 1129227
Shape after dropping: 1128859


In [31]:
# Now we should check if there are any duplicates in the Kaggle dataset with the same name and date
#data_kaggle_cleaned['release_date'] = pd.to_datetime(data_kaggle_cleaned['release_date'], errors='coerce')
duplicates_kaggle_name_date = data_kaggle_cleaned[data_kaggle_cleaned.duplicated(subset=['title', 'release_date'], keep=False)]
duplicates_kaggle_name_date.shape[0]

21249

In [32]:
# In order to enrich our dataset, we have to make both datasets consistent so we are going to drop movies with the same name and date
data_kaggle_cleaned_no_duplicates = data_kaggle_cleaned.drop_duplicates(subset=['title', 'release_date'], keep=False, inplace=False)
data_kaggle_cleaned_no_duplicates.shape[0]

1107610

In [33]:
# So now we want to fill out the missing values in our dataset with the values from kaggle dataset but only for the movies that are in both datasets
# Merge the two datasets on the name and release year

# Here we are going to extract the year from the release date in both datasets
data_no_duplicates['release_year_merge'] = data_no_duplicates['release_date'].str.extract(r'(\d{4})')
data_kaggle_cleaned_no_duplicates['release_year_merge'] = data_kaggle_cleaned_no_duplicates['release_date'].str.extract(r'(\d{4})')

# Save duplicates for later
duplicates = data_no_duplicates[data_no_duplicates.duplicated(subset=['name', 'release_year_merge'], keep=False)]
data_no_duplicates = data_no_duplicates.drop_duplicates(subset=['name', 'release_year_merge'])
data_kaggle_cleaned_no_duplicates = data_kaggle_cleaned_no_duplicates.drop_duplicates(subset=['title', 'release_year_merge'])

# Merge the two datasets on the name and release year
merged_data = pd.merge(
    data_no_duplicates,
    data_kaggle_cleaned_no_duplicates,
    left_on=['name', 'release_year_merge'],
    right_on=['title', 'release_year_merge'],
    suffixes=('_wiki', '_kaggle'),
    how='left'
);
merged_data = merged_data.dropna(subset=['wikipedia_id'])
merged_data.head()

C:\Users\Marija\AppData\Local\Temp\ipykernel_17904\1967789076.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_no_duplicates['release_year_merge'] = data_no_duplicates['release_date'].str.extract(r'(\d{4})')
C:\Users\Marija\AppData\Local\Temp\ipykernel_17904\1967789076.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_kaggle_cleaned_no_duplicates['release_year_merge'] = data_kaggle_cleaned_no_duplicates['release_date'].str.extract(r'(\d{4})')


wikipedia_id freebase_id  \
0        975900   /m/03vyhn   
1       3196793   /m/08yl5d   
2      28463795  /m/0crgdbh   
3       9363483  /m/0285_cd   
4        261236   /m/01mrr1   

                                                    name release_date_wiki  \
0                                         Ghosts of Mars        2001-08-24   
1  Getting Away with Murder: The JonBenét Ramsey Mystery        2000-02-16   
2                                            Brun bitter              1988   
3                                       White Of The Eye              1987   
4                                      A Woman in Flames              1983   

   revenue_wiki  runtime_wiki           languages                 countries  \
0    14010832.0          98.0    English Language  United States of America   
1           NaN          95.0    English Language  United States of America   
2           NaN          83.0  Norwegian Language                    Norway   
3           NaN         110.0    English Language            United Kingdom   
4           NaN         106.0     German Language                   Germany   

                                                                         genres_wiki  \
0  Thriller, Science Fiction, Horror, Adventure, Supernatural, Action, Space western   
1                                     Mystery, Biographical film, Drama, Crime Drama   
2                                                               Crime Fiction, Drama   
3                                  Thriller, Erotic thriller, Psychological thriller   
4                                                                              Drama   

  release_year_merge  ...  \
0               2001  ...   
1               2000  ...   
2               1988  ...   
3               1987  ...   
4               1983  ...   

                                          original_title  \
0                                         Ghosts of Mars   
1  Getting Away with Murder: The JonBenét Ramsey Mystery   
2                                                    NaN   
3                                                    NaN   
4                                    Die flambierte Frau   

                                                                                                                                                                                                  overview  \
0  In 2176, a Martian police unit is sent to pick up a highly dangerous criminal at a remote mining post. Upon arrival, the cops find the post deserted and something far more dangerous than any crimi...   
1  Dramatization of the story behind the murder of young JonBenet Ramsey. The program provides a concise view of the events surrounding the murder case, as pieced together by detectives and reporters...   
2                                                                                                                                                                                                      NaN   
3                                                                                                                                                                                                      NaN   
4  Eva, an upper-class housewife, frustratedly leaves her arrogant husband and decides to enter the call girl business. She lets Yvonne, a prostitute, teach her the basics and both set out for prey t...   

   popularity                       poster_path  \
0      14.189  /i2zztssCIbahGES1fdfWFmDXian.jpg   
1       0.600                               NaN   
2         NaN                               NaN   
3         NaN                               NaN   
4       2.801  /iKeCVR0x8vnRXLtjwjZ7EuVFlo7.jpg   

                             tagline                    genres_kaggle  \
0  Terror is the same on any planet.  Action, Horror, Science Fiction   
1                                NaN                     Drama, Crime   
2                                NaN        

In [34]:
# Now we are going to fill the missing values in the CMU dataset with the values from the kaggle dataset
merged_data['revenue'] = merged_data['revenue_wiki'].fillna(merged_data['revenue_kaggle'])
merged_data['revenue'] = merged_data['revenue'].replace(0, np.nan)
merged_data['runtime'] = merged_data['runtime_wiki'].fillna(merged_data['runtime_kaggle'])
merged_data['genres'] = merged_data['genres_wiki'].fillna(merged_data['genres_kaggle'])
merged_data['countries'] = merged_data['countries'].fillna(merged_data['production_countries'])
merged_data['languages'] = merged_data['languages'].fillna(merged_data['spoken_languages'])


# Now we want to enrich the release date column with days and months
merged_data['release_date'] = np.where(
    merged_data['release_date_kaggle'].str.match(r'^\d{4}-\d{2}-\d{2}$', na=False),
    merged_data['release_date_kaggle'],
    merged_data['release_date_wiki']
)
                      
merged_data['release_date'] = np.where(
    merged_data['release_date_kaggle'].str.match(r'^\d{4}-\d{2}$', na=False) & merged_data['release_date'].str.match(r'^\d{4}$', na=False), #check if wiki has only year
    merged_data['release_date_kaggle'],
    merged_data['release_date']
)

# We are going to drop the columns that we don't need anymore
merged_data.drop(columns=['revenue_wiki', 'revenue_kaggle', 'runtime_wiki', 'runtime_kaggle', 'genres_wiki', 'genres_kaggle', 'production_countries', 'spoken_languages', 
                          'keywords', 'id', 'title', 'vote_average', 'vote_count','status', 'release_date_kaggle', 'release_date_wiki', 'adult','backdrop_path','homepage','imdb_id',
                          'keywords','spoken_languages', 'tagline','production_companies','poster_path', 'original_language','original_title','overview','release_year_merge'], inplace=True)


# add duplicates back to the dataset
duplicates = duplicates.drop(columns=['release_year_merge']) # we don't want this column in our dataset
merged_data = pd.concat([merged_data, duplicates])
#show all the columns
pd.set_option('display.max_columns', None)

In [35]:
# What do we have now?
print("Null values in the merged dataset:")
print(merged_data.isna().sum())

Null values in the merged dataset:
wikipedia_id        0
freebase_id         0
name                0
languages        8874
countries        5806
budget          31699
popularity      31699
revenue         70596
runtime         11734
genres           1797
release_date     6778
dtype: int64


In [36]:
# Have we improved the dataset? Yes, we have filled some of the missing values
# We improved the revenue column with 2744 values
# Release date is improved by adding 21076 more full dates

# Count up the number of movies that have revenue data
print(f"Revenue values count: {merged_data['revenue'].notna().sum()}")

# Count up the number of movies that have full release date
print(f"Full release date count:{merged_data['release_date'].str.match(r'^\d{4}-\d{2}-\d{2}$').sum()}")

# Count up the number of movies that have only year as release date
print(f"Year only release date count:{merged_data['release_date'].str.match(r'^\d{4}$').sum()}")

#count the number of movies with only year and month as release date
print(f"Year and month release date count:{merged_data['release_date'].str.match(r'^\d{4}-\d{2}$').sum()}")

# Count up the number of movies that have missing release date
print(f"Missing release date count:{merged_data['release_date'].isna().sum()}")

Revenue values count: 11032
Full release date count:60448
Year only release date count:13285
Year and month release date count:1117
Missing release date count:6778


In [37]:
merged_data.head()

wikipedia_id freebase_id  \
0        975900   /m/03vyhn   
1       3196793   /m/08yl5d   
2      28463795  /m/0crgdbh   
3       9363483  /m/0285_cd   
4        261236   /m/01mrr1   

                                                    name           languages  \
0                                         Ghosts of Mars    English Language   
1  Getting Away with Murder: The JonBenét Ramsey Mystery    English Language   
2                                            Brun bitter  Norwegian Language   
3                                       White Of The Eye    English Language   
4                                      A Woman in Flames     German Language   

                  countries      budget  popularity     revenue  runtime  \
0  United States of America  28000000.0      14.189  14010832.0     98.0   
1  United States of America         0.0       0.600         NaN     95.0   
2                    Norway         NaN         NaN         NaN     83.0   
3            United Kingdom         NaN         NaN         NaN    110.0   
4                   Germany         0.0       2.801         NaN    106.0   

                                                                              genres  \
0  Thriller, Science Fiction, Horror, Adventure, Supernatural, Action, Space western   
1                                     Mystery, Biographical film, Drama, Crime Drama   
2                                                               Crime Fiction, Drama   
3                                  Thriller, Erotic thriller, Psychological thriller   
4                                                                              Drama   

  release_date  
0   2001-08-24  
1   2000-02-16  
2         1988  
3         1987  
4   1983-05-11

In [38]:
# Sooo now we have more resources for our analysis!
# Let's do a quick analysis for some problematic parts of our new dataset
# We see that budget can have both 0.0 and NaN - we should ensure that every entry has NaN if it doesn't have any meaningful value, just to be consistent
merged_data.loc[merged_data['budget'] == 0.0, 'budget'] = np.nan
merged_data.loc[merged_data['runtime'] == 0.0, 'runtime'] = np.nan #it doesn't make sense to have runtime 0.0

# Now let's check again non-nan values
print("Null values in the merged dataset:")
print(merged_data.isna().sum())

Null values in the merged dataset:
wikipedia_id        0
freebase_id         0
name                0
languages        8874
countries        5806
budget          73338
popularity      31699
revenue         70596
runtime         13676
genres           1797
release_date     6778
dtype: int64


In [39]:
# There might be a lot of genres - let's check
unique_genres = set()

for genre_list in merged_data['genres'].dropna():  # dropna to skip any NaN values
    genres = [genre.strip() for genre in genre_list.split(',')]
    unique_genres.update(genres)

print("Number of unique genres:", len(unique_genres))
print("Unique genres:", unique_genres)

Number of unique genres: 367
Unique genres: {'Tamil cinema', 'Clay animation', 'Music', 'Docudrama', 'Alien Film', 'Legal drama', 'Anime', 'Art film', 'Archives and records', 'Revenge', 'Caper story', 'Heavenly Comedy', 'Interpersonal Relationships', 'News', 'Haunted House Film', 'Comedy Western', 'Glamorized Spy Film', 'Action/Adventure', 'World History', 'Reboot', 'Blaxploitation', 'Sci Fi Pictures original films', 'Gulf War', 'Romance Film', 'Parkour in popular culture', 'Revisionist Fairy Tale', 'Ensemble Film', 'Breakdance', 'Silent film', "Children's Issues", 'Addiction Drama', 'Z movie', 'Romantic thriller', 'Detective', 'Gross out', 'Family Drama', 'Film \\u00e0 clef', 'Fairy tale', 'Doomsday film', 'Kafkaesque', 'Ninja movie', 'Neorealism', 'Natural disaster', "Children's", 'Journalism', 'Anti-war', 'Roadshow theatrical release', 'Movie serial', 'Prison film', 'Escape Film', 'Period Horror', 'Political satire', 'Stand-up comedy', 'Giallo', 'Political cinema', 'Gay Interest', '

In [40]:
# Too much... We should try to find a list of 20 (for example) representable genres, and sort our genres into these groups
genre_categories = {
    'Action': ['Action', 'Action/Adventure', 'Spy', 'Martial Arts Film', 'Superhero'],
    'Adventure': ['Adventure', 'Space opera', 'Swashbuckler films'],
    'Animation': ['Animation', 'Computer Animation', 'Anime', 'Clay animation', 'Stop motion'],
    'Comedy': ['Comedy', 'Comedy-drama', 'Black comedy', 'Slapstick', 'Romantic comedy'],
    'Crime': ['Crime', 'Gangster Film', 'Detective fiction', 'Heist', 'Crime Thriller'],
    'Documentary': ['Documentary', 'Docudrama', 'Rockumentary', 'Essay Film'],
    'Drama': ['Drama', 'Family Drama', 'Inspirational Drama', 'Biographical film', 'Tragedy'],
    'Family': ["Children's", 'Family', "Children's/Family", 'Family Film'],
    'Fantasy': ['Fantasy', 'Sword and sorcery', 'Fairy tale', 'Fantasy Adventure'],
    'Horror': ['Horror', 'Zombie Film', 'Slasher', 'Psychological horror', 'Monster'],
    'Musical': ['Musical', 'Operetta', 'Rockumentary', 'Backstage Musical', 'Jukebox musical'],
    'Mystery': ['Mystery', 'Whodunit', 'Suspense'],
    'Romance': ['Romance', 'Romantic comedy', 'Romantic drama', 'Romantic thriller', 'Chick flick'],
    'Science Fiction': ['Science Fiction', 'Sci-Fi Horror', 'Cyberpunk', 'Alien invasion', 'Time travel'],
    'Thriller': ['Thriller', 'Comedy Thriller', 'Psychological thriller', 'Action Thrillers'],
    'War': ['War', 'War film', 'Cold War', 'Anti-war'],
    'Western': ['Western', 'Spaghetti Western', 'Acid western', 'Epic Western'],
    'Historical': ['Historical', 'Historical Epic', 'Historical drama', 'Period piece'],
    'Superhero': ['Superhero', 'Superhero movie'],
    'Experimental': ['Experimental film', 'Avant-garde', 'Mockumentary', 'Absurdism']
}
def map_genres(genres_string):
    matched_genres = set()  
    
    if pd.notna(genres_string): 
        genres = [genre.strip() for genre in genres_string.split(',')]
        
        for genre in genres:
            for categories, options in genre_categories.items():
                if genre in options:
                    matched_genres.add(categories)
                    break  
    
    return ', '.join(matched_genres) if matched_genres else 'Other'

merged_data['genres'] = merged_data['genres'].apply(map_genres)
print(merged_data['genres'].value_counts())

genres
Drama                                                  15976
Other                                                  11072
Comedy                                                  7238
Documentary                                             4632
Drama, Comedy                                           2474
                                                       ...  
Family, War                                                1
Family, Science Fiction, Action, Superhero                 1
War, Science Fiction, Drama, Horror                        1
Science Fiction, Romance, Drama, Fantasy, Horror           1
Action, Historical, Crime, Drama, Thriller, Mystery        1
Name: count, Length: 2125, dtype: int64


In [41]:
# We do have roughly 12000 'Other' genres, but we are going to leave it like this for now, as it isn't that big of a number for this dataset
# We now want to save this dataset, as we are done for now!

file_path = '../../data/movies_dataset.tsv'
merged_data.to_csv(file_path, sep='\t', index=False) # we are saving it as tsv, because originally it was in tsv format